# Preprocessing

In [22]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("../Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [23]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
#  YOUR CODE GOES HERE
app_df = application_df.drop(columns = ["EIN", "NAME"])
app_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [24]:
# Determine the number of unique values in each column.
#  YOUR CODE GOES HERE
app_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [25]:
# Look at APPLICATION_TYPE value counts for binning
#  YOUR CODE GOES HERE
app_count = application_df['APPLICATION_TYPE'].value_counts()
app_count


T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [26]:
application_df["CLASSIFICATION"].value_counts()[application_df["CLASSIFICATION"].value_counts()>1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [27]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
#  YOUR CODE GOES HERE
application_types_to_replace = list(app_count[app_count<200].index)
# Replace in dataframe
for app in application_types_to_replace:
    app_df['APPLICATION_TYPE'] = app_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
app_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [28]:
# Look at CLASSIFICATION value counts for binning
#  YOUR CODE GOES HERE
app_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [29]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
#  YOUR CODE GOES HERE
classifications_to_replace = list(app_df["CLASSIFICATION"].value_counts()[app_df["CLASSIFICATION"].value_counts()<300].index)

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
Other     1484
C7000      777
Name: CLASSIFICATION, dtype: int64

In [30]:
# Convert categorical data to numeric with `pd.get_dummies`
#  YOUR CODE GOES HERE
app_df = pd.get_dummies(app_df)
app_df.head() 


STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                     1                     0                    0   
1                     0                     0                    1   
2                     0                     0                    0   
3                     0                     0                    1   
4                     0                     0                    1   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    0                    1                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 109 columns]

In [31]:
# Split our preprocessed data into our features and target arrays
#  YOUR CODE GOES HERE
X_features = app_df.drop(["IS_SUCCESSFUL"],axis=1)
y_features = app_df["IS_SUCCESSFUL"]

# Split the preprocessed data into a training and testing dataset
#  YOUR CODE GOES HERE
X_train, X_test, y_train, y_test = train_test_split(X_features, y_features, random_state=1)


In [32]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)
X_test_scaled.shape

(8575, 108)

## Compile, Train and Evaluate the Model

In [33]:
#Trial 2


# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 =  14
hidden_nodes_layer2 = 7
hidden_nodes_layer3 = 4

nn = tf.keras.models.Sequential()

# First hidden layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=46, activation="relu", input_dim=len(X_train_scaled[0])))


# Second hidden layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=34, activation="relu"))
# Output layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))
# Check the structure of the model
nn.summary()

Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 46)                5014      
                                                                 
 dense_7 (Dense)             (None, 34)                1598      
                                                                 
 dense_8 (Dense)             (None, 1)                 35        
                                                                 
Total params: 6,647
Trainable params: 6,647
Non-trainable params: 0
_________________________________________________________________


In [34]:
# Compile the model
#  YOUR CODE GOES HERE
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [35]:
# Train the model
#  YOUR CODE GOES HERE
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5798 - accuracy: 0.7163
Epoch 2/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5540 - accuracy: 0.7302
Epoch 3/100
804/804 [==============================] - 4s 5ms/step - loss: 0.5496 - accuracy: 0.7309
Epoch 4/100
804/804 [==============================] - 4s 5ms/step - loss: 0.5479 - accuracy: 0.7322
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5462 - accuracy: 0.7336
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5452 - accuracy: 0.7345
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5436 - accuracy: 0.7339
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5436 - accuracy: 0.7338
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5428 - accuracy: 0.7357
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5419 - accura

804/804 [==============================] - 1s 2ms/step - loss: 0.5298 - accuracy: 0.7418
Epoch 82/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5298 - accuracy: 0.7410
Epoch 83/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5295 - accuracy: 0.7418
Epoch 84/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5300 - accuracy: 0.7410
Epoch 85/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5299 - accuracy: 0.7423
Epoch 86/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5295 - accuracy: 0.7410
Epoch 87/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5295 - accuracy: 0.7422
Epoch 88/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5294 - accuracy: 0.7422
Epoch 89/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5301 - accuracy: 0.7427
Epoch 90/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5292 - accuracy: 

In [36]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.5600 - accuracy: 0.7263 - 545ms/epoch - 2ms/step
Loss: 0.5599806308746338, Accuracy: 0.7262973785400391


In [37]:
#Trial 3
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
#  YOUR CODE GOES HERE
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 =  14

nn = tf.keras.models.Sequential()

# First hidden layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=46, activation="tanh", input_dim=len(X_train_scaled[0])))


# Second hidden layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=34, activation="tanh"))
# Output layer
#  YOUR CODE GOES HERE
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))
# Check the structure of the model
nn.summary()

Model: "sequential_3"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_9 (Dense)             (None, 46)                5014      
                                                                 
 dense_10 (Dense)            (None, 34)                1598      
                                                                 
 dense_11 (Dense)            (None, 1)                 35        
                                                                 
Total params: 6,647
Trainable params: 6,647
Non-trainable params: 0
_________________________________________________________________


In [38]:
# Compile the model
#  YOUR CODE GOES HERE
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [39]:
# Train the model
#  YOUR CODE GOES HERE
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5720 - accuracy: 0.7199
Epoch 2/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5551 - accuracy: 0.7279
Epoch 3/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5508 - accuracy: 0.7301
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5481 - accuracy: 0.7297
Epoch 5/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5467 - accuracy: 0.7322
Epoch 6/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5450 - accuracy: 0.7331
Epoch 7/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5439 - accuracy: 0.7331
Epoch 8/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5433 - accuracy: 0.7343
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5426 - accuracy: 0.7349
Epoch 10/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5419 - accura

804/804 [==============================] - 2s 2ms/step - loss: 0.5276 - accuracy: 0.7428
Epoch 82/100
804/804 [==============================] - 4s 4ms/step - loss: 0.5276 - accuracy: 0.7434
Epoch 83/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5273 - accuracy: 0.7436
Epoch 84/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5273 - accuracy: 0.7426
Epoch 85/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5274 - accuracy: 0.7432
Epoch 86/100
804/804 [==============================] - 2s 2ms/step - loss: 0.5272 - accuracy: 0.7423
Epoch 87/100
804/804 [==============================] - 2s 3ms/step - loss: 0.5270 - accuracy: 0.7430
Epoch 88/100
804/804 [==============================] - 4s 5ms/step - loss: 0.5270 - accuracy: 0.7422
Epoch 89/100
804/804 [==============================] - 3s 3ms/step - loss: 0.5269 - accuracy: 0.7434
Epoch 90/100
804/804 [==============================] - 3s 4ms/step - loss: 0.5273 - accuracy: 

In [41]:
# Compile the model
#  YOUR CODE GOES HERE
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [42]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5561 - accuracy: 0.7294 - 360ms/epoch - 1ms/step
Loss: 0.5560851097106934, Accuracy: 0.7294460535049438


# Report


### Overview: 
The purpose of creating a deep learning model is to predict if which applicants for funding will be successful. This will help the company optimize its funds to those that are most likely to financially deliver. 


### Results: 
- Target variables: IS_SUCCESSFUL 
- Feature variables: SPECIAL_CONSIDERATIONS, ASK_AMT, APPLICATION_TYPE, INCOME-AMT, STATUS
- Non-target or non-feature variables: USE_CASE, AFFILIATION, CLASSIFICATION, ORGANIZATION
- For my second trial, I 3 hidden layers with 14, 7 and 4 neurons, respectively. In the third trial I only used 1 hidden layer with 14 neurons. In the second trial, I used the relu activation function whereas in the third trial, I used the tanh trial. 
- I was unable to achieve the target model performance. The maximum I reached was 73%. 
- In order to try and increase my model perforance I added 3 hidden layers and added 14, 7, and 4 neurons. In my third trial, I tried to change my activation model from relu to tanh. 

### Summary:
I was unable to achieve the target goal of 75% for this deep learning model. I tried to change the activation model from relu to tanh and I also tried increasing the amount of hidden layers and neurons in each layer. If I were to do another trial or advise someone else, I would remove some variables that seemed irrelevant and increase the number of bins. This more stringent requirements for the data might provide a better final target score. 